sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -474,203 +641,234 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
-        errors.throwIllegalArgument(
-            "%s is not a valid context parameter. Should be one of %s",
-            formatType(paramT), allowedParamTypes);
+      return Parameter.timerParameter(timerDecl);
+
+    } else if (RestrictionTracker.class.isAssignableFrom(rawType)) {
+      errors.checkArgument(
+          !methodContext.getExtraParameters().contains(Parameter.restrictionTracker()),
+          "Multiple %s parameters",
+          RestrictionTracker.class.getSimpleName());
+      return Parameter.restrictionTracker();
+    } else if (State.class.isAssignableFrom(rawType)) {
+      // m.getParameters() is not available until Java 8
+      String id = getStateId(param.getAnnotations());
+      errors.checkArgument(
+          id != null,
+          "parameter of type %s at index %s missing %s annotation",
+          param.getType(),
+          param.getIndex(),
+          DoFn.StateId.class.getSimpleName());
+
+      errors.checkArgument(
+          !methodContext.getStateParameters().containsKey(id),
+          "%s parameter of type %s at index %s duplicates %s(\"%s\") on other parameter",
+          fnClass.getRawType().getName(),
+          param.getType(),
+          param.getIndex(),
+          DoFn.StateId.class.getSimpleName(),
+          id);
+
+      // By static typing this is already a well-formed State subclass
+      TypeDescriptor<? extends State> stateType = (TypeDescriptor<? extends State>) param.getType();
+
+      StateDeclaration stateDecl = globalContext.getStateDeclarations().get(id);
+      errors.checkArgument(
+          stateDecl != null,
+          "parameter of type %s at index %s references undeclared %s \"%s\"",
+          param.getType(),
+          param.getIndex(),
+          DoFn.StateId.class.getSimpleName(),
+          id);
+
+      errors.checkArgument(
+          stateDecl.stateType().equals(stateType),
+          "parameter at index %s has type %s but is a reference to StateId %s of type %s",
+          param.getIndex(),
+          param.getType(),
+          id,
+          stateDecl.stateType());
+
+      errors.checkArgument(
+          stateDecl.field().getDeclaringClass().equals(param.getMethod().getDeclaringClass()),
+          "Method %s has State parameter at index %s for state %s"
+              + " declared in a different class %s."
+              + " State may be referenced only in the class where it is declared.",
+          param.getMethod(),
+          param.getIndex(),
+          id,
+          stateDecl.field().getDeclaringClass().getName());
+
+      return Parameter.stateParameter(stateDecl);
+    } else {
+      List<String> allowedParamTypes =
+          Arrays.asList(
+              formatType(new TypeDescriptor<BoundedWindow>() {}),
+              formatType(new TypeDescriptor<RestrictionTracker<?>>() {}));
+      errors.throwIllegalArgument(
+          "%s is not a valid context parameter. Should be one of %s",
```

Comment:
Should this message be updated to mention that state and timers are also allowed, or is it too early?